torch/distributed/traceable_collectives.py

```diff
@@ -0,0 +1,92 @@
+import torch
+from torch._C import DispatchKey
+from torch._C import _disabled_torch_function_impl
+from torch.utils._pytree import tree_map
+from typing import Any, Optional
+
+"""
+New traceable, functional collectives.
+  compiler: trace these ops with plain-old-data schemas, then choose how to lower them.
+  eager: execute these 'functional' ops which in eager return AsyncCollectiveTensor subclasses,
+         automatically calling .wait() on underlying/hidden async 'work' obj only when fed to
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* How can we make these ops work in eager without manually enabling python_dispatacher mode?
+
+"""
+
+
+class AsyncCollectiveTensor(torch.Tensor):
+    r"""
+    A Tensor subclass that is only used in eager mode, to hold a 'work' object
+    and then wait on it before invoking a real op.
+
+    Usage, from inside functional collective:
+    def functional_collective(input):
+        input = input.clone()
+        mutated_input, work = c10d.{inplace_collective}(input)
+        return AsyncCollectiveTensor(mutated_input, work)
+    """
+    _tensor: torch.Tensor
```

Comment:
This is a kind if interesting design as it implies we have to preallocate the tensor while we wait for the Work to resolve. I think that's always true for the distributed collectives?

Reply:
yea, i didn't check the nccl apis but i am only familiar using collectives as an inplace/mutating op.  do you see this as being a problem?

Reply:
No, it should be OK as long as there aren't any situations where you can't afford to preallocate (i.e. memory constrained)